src/main/java/net/snowflake/client/core/SessionUtil.java

```diff
@@ -334,20 +328,75 @@ static SFLoginOutput openSession(
       }
     }
 
-    preNewSession(loginInput);
+    readCachedTokens(loginInput);
+
+    if (AccessTokenProviderFactory.isEligible(getAuthenticator(loginInput))) {
+      if (loginInput.getOauthAccessToken() != null) { // Access Token cached
+        loginInput.setAuthenticator(AuthenticatorType.OAUTH.name());
+        loginInput.setToken(loginInput.getOauthAccessToken());
+      } else { // Access Token not cached
+        obtainOAuthAccessTokenAndUpdateInput(loginInput);
+      }
+    }
 
     try {
       return newSession(loginInput, connectionPropertiesMap, tracingLevel);
     } catch (SnowflakeReauthenticationRequest ex) {
-      // Id Token expired. We run newSession again with id_token cache cleared
-      logger.debug("ID Token being used has expired. Reauthenticating with ID Token cleared...");
+      if (ex.getErrorCode() == Constants.OAUTH_ACCESS_TOKEN_EXPIRED_GS_CODE) {
+        if (loginInput.getOauthRefreshToken() != null) {
+          refreshOAuthAccessTokenAndUpdateInput(loginInput);
+        } else {
+          loginInput.setAuthenticator(loginInput.getOriginAuthenticator());
+          obtainOAuthAccessTokenAndUpdateInput(loginInput);
+        }
+      }
       return newSession(loginInput, connectionPropertiesMap, tracingLevel);
     }
   }
 
-  private static void preNewSession(SFLoginInput loginInput) throws SFException {
+  private static void obtainOAuthAccessTokenAndUpdateInput(SFLoginInput loginInput)
+      throws SFException {
+    AccessTokenProviderFactory accessTokenProviderFactory =
+        new AccessTokenProviderFactory(
+            new SessionUtilExternalBrowser.DefaultAuthExternalBrowserHandlers(),
+            (int) loginInput.getBrowserResponseTimeout().getSeconds());
+    AccessTokenProvider accessTokenProvider =
+        accessTokenProviderFactory.createAccessTokenProvider(
+            getAuthenticator(loginInput), loginInput);
+    TokenResponseDTO tokenResponse = accessTokenProvider.getAccessToken(loginInput);
+    loginInput.setAuthenticator(AuthenticatorType.OAUTH.name());
+    loginInput.setToken(tokenResponse.getAccessToken());
+    loginInput.setOauthAccessToken(tokenResponse.getAccessToken());
```

Comment:
why do we pass access token twice?

Reply:
`Token` field is generic token field which is related to /login-request body and `oauthToken` on the other hand is a field that is being used to store only oauth access token and to interact with token cache